streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -0,0 +1,344 @@
+    private final Serde<K> keySerde;
+    private final Serde<V> valueSerde;
+    private final String metricScope;
+    private StateSerdes<K, V> serdes;
+    private InternalProcessorContext context;
+    private Sensor expiredRecordSensor;
+
+    private final long retentionPeriod;
+    private final long windowSize;
+    private final long gracePeriod;
+
+    private final NavigableMap<Long, NavigableMap<K, V>> segmentMap;
+
+    private volatile boolean open = false;
+
+    public InMemoryWindowStore(final String name,
+        final Serde<K> keySerde,
+        final Serde<V> valueSerde,
+        final long retentionPeriod,
+        final long windowSize,
+        final long gracePeriod,
+        final String metricScope) {
+        this.name = name;
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.retentionPeriod = retentionPeriod;
+        this.windowSize = windowSize;
+        this.gracePeriod = gracePeriod;
+        this.metricScope = metricScope;
+
+        this.segmentMap = new TreeMap<>();
+    }
+
+    @Override
+    public String name() {
+        return this.name;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void init(final ProcessorContext context, final StateStore root) {
+        this.context = (InternalProcessorContext) context;
+
+        // construct the serde
+        this.serdes = new StateSerdes<>(
+            ProcessorStateManager.storeChangelogTopic(context.applicationId(), name),
+            keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
+            valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+
+        final StreamsMetricsImpl metrics = this.context.metrics();
+        final String taskName = context.taskId().toString();
+        expiredRecordSensor = metrics.storeLevelSensor(
+            taskName,
+            name(),
+            "expired-window-record-drop",
+            Sensor.RecordingLevel.INFO
+        );
+        addInvocationRateAndCount(
+            expiredRecordSensor,
+            "stream-" + metricScope + "-metrics",
+            metrics.tagMap("task-id", taskName, metricScope + "-id", name()),
+            "expired-window-record-drop"
+        );
+
+        if (root != null) {
+            context.register(root, (RecordBatchingStateRestoreCallback) records -> {
+                for (final ConsumerRecord<byte[], byte[]> record : records) {
+                    put(extractStoreKey(record.key(), serdes), serdes.valueFrom(record.value()), extractStoreTimestamp(record.key()));
+                }
+            });
+        }
+        this.open = true;
+    }
+
+    @Override
+    public void put(final K key, final V value) {
+        put(key, value, context.timestamp());
+    }
+
+    @Override
+    public void put(final K key, final V value, final long windowStartTimestamp) {
+        removeExpiredSegments();
+        if (windowStartTimestamp <= this.context.streamTime() - this.gracePeriod) {
```

Comment:
It shouldn't be necessary for the store to enforce a grace period.

The API is a little muddled because we're transitioning from, and still compatible with, an older strategy, but the current idea is:
* windowed operators (like windowed aggregations) enforce grace period, which they receive via the window definition.
* window stores have a concept of retention period
* the DSL enforces that the retention period of a window store is great enough to accommodate the window definition's full lifecycle (window size + grace period).

Thus, I *think* the right thing to do is actually check that the record is within the retention period right here, and remove `gracePeriod` from this class entirely.... What do you think?

Reply:
Makes sense to me. Will remove grace period from the store implementation and only check against retention period here